### PR TITLE
P2P / Engine API Changes for Dencun

### PR DESCRIPTION
**Pull request by danyalprout**
**Description**

P2P / serialization / engine API changes for Dencun. Also updated the admin API to receive an ExecutionPayloadEnvelope instead of an ExecutionPayload

#### Remaining Dencun Todos

- [x] Update to a release version of op-geth. Requires this [PR](https://github.com/ethereum-optimism/op-geth/pull/214)
- [x] Decide how to handle sync client changes.
- [ ] Modify `op-ha` to migrate to the new admin API
- [ ] Investigate whether to migrate `op-wheel` to V3 engine APIs
- [ ] Add some more tests for Ecotone specific blocks (parent beacon block root != nil)

**Tests**

Added tests around:

* SSZ
* Gossip
* V3 migration

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -219,7 +219,8 @@ require (
 	rsc.io/tmplfunc v0.0.3 // indirect
 )
 
-replace github.com/ethereum/go-ethereum v1.13.5 => github.com/ethereum-optimism/op-geth v1.101305.1-rc.1.0.20240103210259-c7871bc4454f
+//replace github.com/ethereum/go-ethereum v1.13.5 => github.com/ethereum-optimism/op-geth v1.101305.0-rc.2.0.20231220194825-94ca8ca83871
+replace github.com/ethereum/go-ethereum v1.13.5 => /Users/danyal/code/ext/op-geth
```
Comment (danyalprout) [inline go.mod]: Requires this to be merged: https://github.com/ethereum-optimism/op-geth/pull/214

File: op-e2e/e2eutils/setup.go [hunk 2]
```diff
@@ -195,3 +195,10 @@ func CanyonTimeOffset() *hexutil.Uint64 {
 	}
 	return nil
 }
+func EcotoneTimeOffset() *hexutil.Uint64 {
```
Comment (tynes) [inline op-e2e/e2eutils/setup.go:199]: nit: no newline

**Comment by protolambda** [comment]
tess pass locally, not sure why CI is timing out. I'll review this, get it onto the dencun branch, and rebase on develop. Should confirm that there's nothing missing. Probably also need the L1-info changes on develop, for this to rebase onto, since Ecotone needs to be stable.

**Comment by protolambda** [comment]
Waiting for #8786 to get on develop, then rebasing the `dencun` branch, and then rebasing this on top of that new `dencun` branch. The remaining TODO issues in this PR are less critical, and we should probably split off into a separate PR, to keep this one manageable and reviewable.

**Comment by protolambda** [comment]
Rebased on dencun branch, resolved some big merge conflicts with the engine-controller refactor, squashed to make the commits more manageable with rebasing, and fixed gossip publishing side.

**Comment by protolambda** [comment]
I'm merging this into the `dencun` branch now, so we can have 1 view of Dencun work again, and get a working engine API for testing of the other dencun changes.

**Pull request merged**